crates/forge/bin/cmd/clone.rs

```diff
@@ -0,0 +1,641 @@
+        }
+        let entry = entry?;
+        let folder_name = entry.file_name().to_string_lossy().to_string();
+        // special handling for contracts and src directories: we flatten them.
+        if folder_name.as_str() == "contracts" || folder_name.as_str() == "src" {
+            // move all sub folders in contracts to src
+            for e in read_dir(entry.path())? {
+                let e = e?;
+                let dest = src_dir.join(e.file_name());
+                std::fs::rename(e.path(), &dest)?;
+                remappings.push(Remapping {
+                    context: None,
+                    name: format!("{}/{}", folder_name, e.file_name().to_string_lossy()),
+                    path: dest.to_string_lossy().to_string(),
+                });
+            }
+        } else {
+            // move the other folders to src
+            let dest = src_dir.join(entry.file_name());
+            std::fs::rename(entry.path(), &dest)?;
+            remappings.push(Remapping {
+                context: None,
+                name: entry.file_name().to_string_lossy().to_string(),
+                path: dest.to_string_lossy().to_string(),
+            });
+        }
+    }
+
+    // remove the temporary directory
+    std::fs::remove_dir_all(tmp_dump_dir)?;
+
+    Ok(remappings.into_iter().map(|r| r.into_relative(root)).collect())
+}
+
+/// Compile the project in the root directory, and return the compilation result.
+pub fn compile_project(root: &PathBuf) -> Result<ProjectCompileOutput> {
+    std::env::set_current_dir(root)?;
+    let mut config = Config::load_with_root(root);
+    config.extra_output.push(ContractOutputSelection::StorageLayout);
+    let project = config.project()?;
+    let compiler = ProjectCompiler::new();
+    compiler.compile(&project)
+}
+
+/// Find the artifact of the contract with the specified name.
+/// This function returns the path to the source file and the artifact.
+pub fn find_main_contract<'a>(
+    compile_output: &'a ProjectCompileOutput,
+    contract: &str,
+) -> Result<(PathBuf, &'a ConfigurableContractArtifact)> {
+    // XXX (ZZ): it is possible that we have multiple contracts with the same name
+    // in different files, we should handle this case in the future
+    for (f, c, a) in compile_output.artifacts_with_files() {
+        if contract == c {
+            return Ok((PathBuf::from(f), a));
+        }
+    }
+    Err(eyre::eyre!("contract not found"))
+}
+
+/// EtherscanClient is a trait that defines the methods to interact with Etherscan.
+/// It is defined as a wrapper of the `foundry_block_explorers::Client` to allow mocking.
+#[automock]
+pub(crate) trait EtherscanClient {
+    async fn contract_source_code(
+        &self,
+        address: Address,
+    ) -> std::result::Result<ContractMetadata, EtherscanError>;
+    async fn contract_creation_data(
+        &self,
+        address: Address,
+    ) -> std::result::Result<ContractCreationData, EtherscanError>;
+}
```

Comment:
I use `#[cfg_attr(test, automock)` here, but `EtherscanClient` trait itself cannot be moved to test mod since it is used in the `run_with_client` function. Please see if this is appropriate or not.